src/export/bops/index.ts

```diff
@@ -0,0 +1,534 @@
+import { GraphQLClient } from 'graphql-request';
+import isEmpty from "lodash.isempty";
+import isNil from "lodash.isnil";
+import { getSessionById } from '../../session';
+import { getLatestFlowGraph } from '../../flow';
+import { BOPSFullPayload, QuestionMetaData, ResponseMetaData, Response, QuestionAndResponses, USER_ROLES, FileTag } from './model';
+import { Breadcrumbs, ComponentType, Flag, FlowGraph, GOV_PAY_PASSPORT_KEY, GovUKPayment, Passport, Value } from '../../types';
+
+export async function generateBOPSPayload(client: GraphQLClient, sessionId: string): Promise<BOPSFullPayload> {
+  const session = await getSessionById(client, sessionId);
+  const flow = await getLatestFlowGraph(client, session.flowId);
+  if (!flow) throw Error(`Cannot get flow ${session.flowId}, therefore cannot generate BOPS payload.`);
+
+  const payload = getBOPSParams({ 
+    breadcrumbs: session.data.breadcrumbs, 
+    flow: flow, 
+    passport: session.data.passport, 
+    sessionId: session.id, 
+    flowName: "TODO",
+  });
+  return payload;
+}
+
+export const bopsDictionary = {
+  // applicant or agent details can be provided via TextInput(plural) or ContactInput component
+  applicant_first_name: "applicant.name.first",
+  applicant_last_name: "applicant.name.last",
+  applicant_phone: "applicant.phone.primary",
+  applicant_email: "applicant.email",
+
+  agent_first_name: "applicant.agent.name.first",
+  agent_last_name: "applicant.agent.name.last",
+  agent_phone: "applicant.agent.phone.primary",
+  agent_email: "applicant.agent.email",
+
+  description: "proposal.description",
+};
+
+function isTypeForBopsPayload(type?: ComponentType) {
+  if (!type) return false;
+
+  switch (type) {
+    case ComponentType.Calculate:
+    case ComponentType.Confirmation:
+    case ComponentType.Content:
+    case ComponentType.DrawBoundary:
+    case ComponentType.ExternalPortal:
+    case ComponentType.FileUpload:
+    case ComponentType.MultipleFileUpload:
+    case ComponentType.Filter:
+    case ComponentType.FindProperty:
+    case ComponentType.Flow:
+    case ComponentType.InternalPortal:
+    case ComponentType.Notice:
+    case ComponentType.Pay:
+    case ComponentType.PlanningConstraints:
+    case ComponentType.PropertyInformation:
+    case ComponentType.Answer:
+    case ComponentType.Result:
+    case ComponentType.Review:
+    case ComponentType.Section:
+    case ComponentType.Send:
+    case ComponentType.SetValue:
+    case ComponentType.TaskList:
+      return false;
+
+    case ComponentType.AddressInput:
+    case ComponentType.Checklist:
+    case ComponentType.DateInput:
+    case ComponentType.NumberInput:
+    case ComponentType.Question:
+    case ComponentType.TextInput:
+    case ComponentType.ContactInput:
+      return true;
+
+    default:
+      throw new Error(`Unhandled type: ${type}`);
```

Comment:
I saw a clever trick for getting typescript to show an error when a case statement isn't exhaustive (e.g. when we add a new ComponentType):
![image](https://user-images.githubusercontent.com/109954107/234529210-7bd4a2e5-939e-476e-a94d-cc7f88172a0d.png)
https://medium.com/technogise/type-safe-and-exhaustive-switch-statements-aka-pattern-matching-in-typescript-e3febd433a7a

Reply:
thanks for pointing this out! we do something similar in planx-new in a couple places using an `exhaustiveCheck` function, which was actually improperly implemented in the BOPS file over there so I didn't copy it over and hadn't updated it yet - but added back properly now :+1: